AIO/src/org/aio/activities/quests/SheepShearer.java

```diff
@@ -0,0 +1,150 @@
+                    } else {
+                        getItemsNeeded();
+                    }
+                    break;
+                case 21:
+                    log("Quest is complete");
+                    isComplete = true;
+                    break;
+                default:
+                    log("Unknown progress config value: " + getProgress());
+                    setFailed();
+                    break;
+            }
+        }
+    }
+
+    private boolean hasRequiredItems() {
+        return getInventory().getAmount("Ball of wool") > 19;
+    }
+
+    private void getItemsNeeded() throws InterruptedException {
+        if (!getInventory().contains("Shears")) {
+            getGroundItem(FARMER, "Shears");
+        } else if (!getInventory().contains("Ball of wool") && getInventory().getAmount("Wool") < 20) {
+            getItemFromNPC(SHEEP, "Shear");
+        } else if (getInventory().getAmount("Ball of wool") < 20) {
+            useObject(SPINNER, "Spin");
+        }
+    }
+
+    private void getItemFromNPC(Area place, String interaction) throws InterruptedException {
+        if (place.contains(myPlayer())) {
+            NPC npc = getNpcs().closest(n -> !n.hasAction("Talk-to") && n.hasAction(interaction) && place.contains(n));
+            if (npc != null && npc.interact(interaction)) {
+                Sleep.sleepUntil(() -> myPlayer().getAnimation() == 893, 15000);
+                Sleep.sleepUntil(() -> !myPlayer().isAnimating(), 10000);
+            }
+        } else {
+            getWalking().webWalk(place);
+        }
+    }
+
+    private void useObject(Area place, String objectAction) throws InterruptedException {
+        if (place.contains(myPlayer())) {
+            RS2Object object = getObjects().closest(n -> n.hasAction(objectAction));
+            if (object != null && object.interact(objectAction)) {
+                Sleep.sleepUntil(() -> getWidgets().getWidgetContainingText("What would you like to spin?") != null, 8000);
+
+                if (makeAllInterface.isMakeAllScreenOpen() && makeAllInterface.makeAll()) {
+                    Sleep.sleepUntil(() -> !getInventory().contains("Wool"), 180000, 1000);
+                }
+            }
+        } else {
+            getWalking().webWalk(place);
+        }
+    }
+
+    private void talkToFarmer() throws InterruptedException {
+        NPC farmer = getNpcs().closest("Fred the Farmer");
+
+        if (!FARMER.contains(myPosition())) {
+            getWalking().webWalk(FARMER);
+        } else if (!getDialogues().inDialogue() || !myPlayer().isInteracting(farmer)) {
+            if (farmer.interact("Talk-to")) {
+                Sleep.sleepUntil(() -> getDialogues().inDialogue() && myPlayer().isInteracting(farmer), 5000);
+            }
+        } else {
+            getDialogues().completeDialogue(FARMER_OPTIONS);
+        }
+    }
+
+    private void getGroundItem(Area place, String itemName) throws InterruptedException {
```

Comment:
These are great generic functions. I'd love to see them put in a `AIO/src/org/aio/activities/quests/QuestUtils.java` helper class!

Reply:
I agree, I use a lot of functions mutliple times in the quest scripts I make. Will look into this in the weekend.